cmd/osbuild-worker-executor/handler_build.go

```diff
@@ -39,63 +41,96 @@ func (wf *writeFlusher) Write(p []byte) (n int, err error) {
 	return n, err
 }
 
-func runOsbuild(logger *logrus.Logger, buildDir string, control *controlJSON, output io.Writer) (string, error) {
-	flusher, ok := output.(http.Flusher)
-	if !ok {
-		return "", fmt.Errorf("cannot stream the output")
-	}
-	// stream output over http
-	wf := writeFlusher{w: output, flusher: flusher}
+func runOSBuild(logger *logrus.Logger, buildDir string, control *controlJSON, output io.Writer) (string, error) {
 	// note that the "filepath.Clean()" is here for gosec:G304
 	logfPath := filepath.Clean(filepath.Join(buildDir, "build.log"))
-	// and also write to our internal log
 	logf, err := os.Create(logfPath)
 	if err != nil {
 		return "", fmt.Errorf("cannot create log file: %v", err)
 	}
 	defer logf.Close()
 
+	manifest, err := os.ReadFile(filepath.Join(buildDir, "manifest.json"))
+	if err != nil {
+		return "", fmt.Errorf("cannot read manifest file: %w", err)
+	}
+
 	// use multi writer to get same output for stream and log
-	mw := io.MultiWriter(&wf, logf)
 	outputDir := filepath.Join(buildDir, "output")
 	storeDir := filepath.Join(buildDir, "osbuild-store")
-	cmd := exec.Command(osbuildBinary)
-	cmd.Stdout = mw
-	cmd.Stderr = mw
-	for _, exp := range control.Exports {
-		cmd.Args = append(cmd.Args, []string{"--export", exp}...)
+
+	// MonitorFile needs an *os.File:
+	// MonitorFile -> pipe -> bufio reader -> writeFlusher
+	rPipe, wPipe, err := os.Pipe()
+	if err != nil {
+		return "", fmt.Errorf("cannot create pipe for monitor file: %w", err)
 	}
-	cmd.Env = append(cmd.Env, control.Environments...)
-	cmd.Args = append(cmd.Args, []string{"--output-dir", outputDir}...)
-	cmd.Args = append(cmd.Args, []string{"--store", storeDir}...)
-	cmd.Args = append(cmd.Args, "--json")
-	cmd.Args = append(cmd.Args, filepath.Join(buildDir, "manifest.json"))
+	defer rPipe.Close()
+	defer wPipe.Close()
+
+	cmd := osbuild.NewOSBuildCmd(manifest, &osbuild.OSBuildOptions{
+		StoreDir:    storeDir,
+		OutputDir:   outputDir,
+		Exports:     control.Exports,
+		ExtraEnv:    control.Environments,
+		BuildLog:    logf,
+		Stdout:      os.Stdout,
+		Monitor:     osbuild.MonitorJSONSeq,
+		MonitorFile: wPipe,
+	})
 	if err := cmd.Start(); err != nil {
 		return "", err
 	}
+	flusher, ok := output.(http.Flusher)
```

Comment:
You could create a embedded interface `WriterFlusher` and make it part of the contract. This is acceptable, but raises a red flag.